label_studio/fsm/state_manager.py

```diff
@@ -0,0 +1,385 @@
+    ) -> bool:
+        """
+        Perform state transition with audit trail.
+
+        Uses INSERT-only approach for maximum performance:
+        - No UPDATE operations or row locks
+        - Complete audit trail by design
+        - Basic cache update for consistency
+
+        Args:
+            entity: The entity to transition
+            new_state: Target state
+            transition_name: Name of transition method (for audit)
+            user: User triggering the transition
+            context: Additional context data
+            reason: Human-readable reason for transition
+
+        Returns:
+            True if transition succeeded, False otherwise
+
+        Raises:
+            InvalidTransitionError: If transition is not valid
+            StateManagerError: If transition fails
+
+        Example:
+            success = StateManager.transition_state(
+                entity=task,
+                new_state='IN_PROGRESS',
+                transition_name='start_annotation',
+                user=request.user,
+                context={'assignment_id': assignment.id},
+                reason='User started annotation work'
+            )
+        """
+        state_model = get_state_model_for_entity(entity)
+        if not state_model:
+            raise StateManagerError(f'No state model found for {entity._meta.model_name}')
+
+        current_state = cls.get_current_state(entity)
+
+        logger.info(
+            f'Transitioning {entity._meta.label_lower} {entity.pk}: '
+            f'{current_state} → {new_state} (transition: {transition_name})'
+        )
+
+        try:
+            with transaction.atomic():
+                # INSERT-only approach - no UPDATE operations needed
+                # Get denormalized fields from the state model itself
+                denormalized_fields = {}
+                if hasattr(state_model, 'get_denormalized_fields'):
+                    denormalized_fields = state_model.get_denormalized_fields(entity)
+
+                # Get organization from user's active organization
+                organization_id = (
+                    user.active_organization.id
+                    if user and hasattr(user, 'active_organization') and user.active_organization
+                    else None
+                )
+
+                new_state_record = state_model.objects.create(
+                    **{entity._meta.model_name: entity},
+                    state=new_state,
+                    previous_state=current_state,
+                    transition_name=transition_name,
+                    triggered_by=user,
+                    context_data=context or {},
+                    reason=reason,
+                    organization_id=organization_id,
+                    **denormalized_fields,
+                )
+
+                # Update cache with new state
+                cache_key = cls.get_cache_key(entity)
+                cache.set(cache_key, new_state, cls.CACHE_TTL)
```

Comment:
I wonder whether it's safer to write cache in the callback upon a successful commit, like
```
    transaction.on_commit(
        lambda: cache.set(cache_key, new_state, cls.CACHE_TTL)
    )
```